src/compiler/checker.ts

```diff
@@ -9567,16 +9567,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
         if (!type) {
             let types: Type[] | undefined;
+            let declarations: Declaration[] | undefined;
             if (symbol.declarations) {
                 let jsdocType: Type | undefined;
                 for (const declaration of symbol.declarations) {
+                    if (isInJSFile(declaration) && hasSyntacticModifier(declaration, ModifierFlags.Ambient)) {

```

Comment:
@sandersn

Reply:
Yes, a property declaration with a type annotation should always skip inference, even if there's a modifier that's illegal in JS. (although there should still be two "doesn't work in JS" errors: one for the modifier and one for the type annotation)

From the PR changes, it looks like the bug is that we somehow miss that annotation when the ambient property declaration is after the constructor, and then that declaration causes JS-property type inference to crash. I don't think the inference code needs to change; I think it should not be called at all.

Reply:
> `it should not be called at all.`

Is it acceptable to avoid adding a _declaration_ for invalid js in a _binder_?

Reply:
No, it should still be bound as if it were TS. Is checking for a type annotation only happening on the value declaration? If so, can `declare x: number` be a value declaration? If it can, but JS assignment declarations are also value declarations, then the binder would need to change to set the ambient declaration as the value declaration even if it comes after JS assignment declarations.